File: packages/feedback/src/default.ts
import { createFeedbackResolver } from "@composable-model-graph/core";
import type { FeedbackResolver } from "@composable-model-graph/core";

/** Options for {@link defaultFeedbackResolver}. */
export interface DefaultFeedbackResolverOptions {
  id?: string;
  name?: string;
}

/**
 * Map an evaluation status to a feedback action:
 *
 *   pass    -> accept
 *   partial -> adjust
 *   fail    -> retry
 *   unknown -> custom (inspect)
 *
 * A run with no evaluation is treated as `unknown`.
 */
export function defaultFeedbackResolver<I = unknown, O = unknown>(
  options: DefaultFeedbackResolverOptions = {},
): FeedbackResolver<I, O> {
  return createFeedbackResolver<I, O>({
    id: options.id ?? "default-feedback",
    name: options.name ?? "Default Feedback",
    resolve: (run) => {
      const status = run.evaluation?.status ?? "unknown";
      switch (status) {
        case "pass":
          return { kind: "accept", reason: "evaluation passed" };
        case "partial":
          return { kind: "adjust", reason: "evaluation partial" };
        case "fail":
          return { kind: "retry", reason: "evaluation failed" };
        default:
          return {
            kind: "custom",
            reason: "inspect",
            signal: { action: "inspect" },
          };
      }
    },
  });
}

File: python/src/composable_model_graph/feedback/default.py
"""Map an evaluation status to a feedback action.

    pass    -> accept
    partial -> adjust
    fail    -> retry
    unknown -> custom (inspect)

A run with no evaluation is treated as `unknown`.

(Python parity of typescript feedback/default.ts.)
"""

from __future__ import annotations

from typing import Any, Optional

from ..core import FeedbackAction, FeedbackResolver, create_feedback_resolver


def default_feedback_resolver(
    id: Optional[str] = None,
    name: Optional[str] = None,
) -> FeedbackResolver:
    def resolve(run: Any, context: Any) -> FeedbackAction:
        status = run.evaluation.status if run.evaluation else "unknown"
        if status == "pass":
            return FeedbackAction(kind="accept", reason="evaluation passed")
        if status == "partial":
            return FeedbackAction(kind="adjust", reason="evaluation partial")
        if status == "fail":
            return FeedbackAction(kind="retry", reason="evaluation failed")
        return FeedbackAction(
            kind="custom", reason="inspect", signal={"action": "inspect"}
        )

    return create_feedback_resolver(
        id=id if id is not None else "default-feedback",
        name=name if name is not None else "Default Feedback",
        resolve=resolve,
    )
